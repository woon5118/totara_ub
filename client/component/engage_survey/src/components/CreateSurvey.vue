<!--
  This file is part of Totara Enterprise Extensions.

  Copyright (C) 2020 onwards Totara Learning Solutions LTD

  Totara Enterprise Extensions is provided only to Totara
  Learning Solutions LTD's customers and partners, pursuant to
  the terms and conditions of a separate agreement with Totara
  Learning Solutions LTD or its affiliate.

  If you do not have an agreement with Totara Learning Solutions
  LTD, you may not access, use, modify, or distribute this software.
  Please contact [licensing@totaralearning.com] for more information.

  @author Johannes Cilliers <johannes.cilliers@totaralearning.com>
  @module engage_survey
-->

<template>
  <div class="tui-engageCreateSurvey">
    <SurveyForm
      v-show="stage === 0"
      :survey="survey"
      @next="next"
      @cancel="$emit('cancel')"
      @unsaved-changes="$emit('unsaved-changes')"
    />
    <AccessForm
      v-show="stage === 1"
      item-id="0"
      component="engage_survey"
      :show-back="true"
      :submitting="submitting"
      :selected-access="containerValues.access || defaultAccess"
      :private-disabled="privateDisabled"
      :restricted-disabled="restrictedDisabled"
      :container="container"
      @done="done"
      @back="back"
      @cancel="$emit('cancel')"
    />
  </div>
</template>

<script>
import SurveyForm from 'engage_survey/components/form/SurveyForm';
import AccessForm from 'totara_engage/components/form/AccessForm';
import createSurvey from 'engage_survey/graphql/create_survey';
import { AccessManager } from 'totara_engage/index';
import { notify } from 'tui/notifications';

// Mixins
import ContainerMixin from 'totara_engage/mixins/container_mixin';

export default {
  components: {
    SurveyForm,
    AccessForm,
  },

  mixins: [ContainerMixin],

  props: {
    showNotification: {
      type: Boolean,
    },
  },

  data() {
    return {
      stage: 0,
      maxStage: 1,
      survey: {
        question: '',
        type: '',
        options: [],
      },
      submitting: false,
      defaultAccess: 'PRIVATE',
    };
  },

  computed: {
    privateDisabled() {
      return this.containerValues.access
        ? !AccessManager.isPrivate(this.containerValues.access)
        : false;
    },
    restrictedDisabled() {
      return this.containerValues.access
        ? AccessManager.isPublic(this.containerValues.access)
        : false;
    },
  },

  methods: {
    /**
     * @param {String}          question
     * @param {Number|String}   type
     * @param {Array}           options
     */
    next({ question, type, options }) {
      if (this.stage < this.maxStage) {
        this.stage += 1;
      }

      this.survey.question = question;
      this.survey.type = type;
      this.survey.options = options;

      this.$emit('change-title', this.stage);
    },

    back() {
      if (this.stage > 0) {
        this.stage -= 1;
      }

      this.$emit('change-title', this.stage);
    },

    /**
     * @param {String} access
     * @param {Array} topics
     * @param {Array} shares
     */
    done({ access, topics, shares }) {
      if (this.submitting) {
        return;
      }
      this.submitting = true;

      this.$apollo
        .mutate({
          mutation: createSurvey,
          refetchQueries: [
            'totara_engage_contribution_cards',
            'container_workspace_contribution_cards',
            'container_workspace_shared_cards',
          ],
          variables: {
            // TODO: replace timeexpired with the time selected from the date component
            timeexpired: null,
            questions: [
              {
                value: this.survey.question,
                answertype: this.survey.type,
                options: this.survey.options.map(option => option.text),
              },
            ],
            access: access,
            topics: topics.map(topic => topic.id),
            shares: shares,
          },
          update: (cache, { data: { survey } }) => {
            this.$emit('done', { resourceId: survey.resource.id });
          },
        })
        .then(({ data: { survey } }) => {
          if (survey && this.showNotification && !this.container) {
            notify({
              message: this.$str('created', 'engage_survey'),
              type: 'success',
            });
          }
          this.$emit('cancel');
        })
        .finally(() => (this.submitting = false));
    },
  },
};
</script>

<lang-strings>
{
  "engage_survey": [
    "created"
  ]
}
</lang-strings>

<style lang="scss">
.tui-engageCreateSurvey {
  display: flex;
  flex: 1;
  flex-direction: column;
}
</style>
