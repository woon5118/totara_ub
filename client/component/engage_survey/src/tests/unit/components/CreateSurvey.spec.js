/**
 * This file is part of Totara Enterprise Extensions.
 *
 * Copyright (C) 2020 onwards Totara Learning Solutions LTD
 *
 * Totara Enterprise Extensions is provided only to Totara
 * Learning Solutions LTD's customers and partners, pursuant to
 * the terms and conditions of a separate agreement with Totara
 * Learning Solutions LTD or its affiliate.
 *
 * If you do not have an agreement with Totara Learning Solutions
 * LTD, you may not access, use, modify, or distribute this software.
 * Please contact [licensing@totaralearning.com] for more information.
 *
 * @author Kian Nguyen <kian.nguyen@totaralearning.com>
 * @module engage_survey
 */

import CreateSurvey from 'engage_survey/components/CreateSurvey';
import { shallowMount } from '@vue/test-utils';

describe('engage_survey/components/CreateSurvey.vue', function() {
  let wrapper = null;

  beforeAll(function() {
    wrapper = shallowMount(CreateSurvey, {
      propsData: {
        itemId: 1,
        component: 'engage_survey',
      },
      mocks: {
        $str(id, component) {
          return `${id}, ${component}`;
        },
      },
    });
  });

  it('Checks snapshot', function() {
    expect(wrapper.element).toMatchSnapshot();
  });
});
