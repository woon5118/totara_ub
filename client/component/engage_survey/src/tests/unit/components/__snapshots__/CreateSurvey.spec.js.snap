// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`engage_survey/components/CreateSurvey.vue Checks snapshot 1`] = `
<div
  class="tui-engageCreateSurvey"
  component="engage_survey"
  itemid="1"
>
  <surveyform-stub
    buttoncontent="next, core"
    showbuttonright="true"
    survey="[object Object]"
  />
   
  <accessform-stub
    component="engage_survey"
    donebuttontext="done, totara_engage"
    itemid="0"
    selectedaccess="PRIVATE"
    selectedoptions="[object Object]"
    showback="true"
    style="display: none;"
  />
</div>
`;
